trace/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,161 @@
+package elasticsearch
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"runtime"
+	"strings"
+
+	"github.com/beatlabs/patron/trace"
+	"github.com/elastic/go-elasticsearch/v6"
+	"github.com/elastic/go-elasticsearch/v6/esapi"
+	"github.com/elastic/go-elasticsearch/v6/estransport"
+	"github.com/opentracing/opentracing-go"
+)
+
+const (
+	// If no addresses are passed, hit localhost by default
+	defaultURL = "http://localhost:9200"
+
+	// span tag that denotes the host that responded to a query
+	respondentTag = "respondent"
+)
+
+type tracingInfo struct {
+	user  string
+	hosts []string
+}
+
+func (t *tracingInfo) startSpan(
+	opName string,
+	req *http.Request,
+) (opentracing.Span, error) {
+
+	if req == nil {
+		return nil, fmt.Errorf("request is empty")
+	}
+
+	uri := req.URL.RequestURI()
+	method := req.Method
+
+	var body string
+	if req.Body != nil {
+		rawBody, err := ioutil.ReadAll(req.Body)
+		if err == nil {
+			body = string(rawBody)
+		}
+	}
+
+	return trace.ElasticsearchSpan(req.Context(), opName, "go-elasticsearch", t.user, uri, method, body, t.hosts), nil
+}
+
+func endSpan(sp opentracing.Span, rsp *http.Response) {
+	// Fetch the host that responded and inject it to the span
+	sp.SetTag(respondentTag, rsp.Request.URL.Host)
+
+	trace.FinishHTTPSpan(sp, rsp.StatusCode)
+}
+
+// transportClient is ES transport struct wrapped with tracing info
+type transportClient struct {
+	client *estransport.Client
+	tracingInfo
+}
+
+func (c *transportClient) Perform(req *http.Request) (*http.Response, error) {
+	// First try to retrieve the name of the caller to pass opName
+	pc, _, _, ok := runtime.Caller(1)
+	opInfo := runtime.FuncForPC(pc)
+	var opName string
+	if ok && opInfo != nil {
+		opName = opInfo.Name()
+	} else {
+		opName = "Transport.Perform"
+	}
+
+	sp, err := c.startSpan(opName, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hit upstream perform
+	rsp, err := c.client.Perform(req)
+	if err != nil || rsp == nil {
+		trace.SpanError(sp)
+		return rsp, err
+	}
+	endSpan(sp, rsp)
+	return rsp, nil
+}
+
+// Config is a wrapper for elasticsearch.Config
```

Comment:
At first, I thought it would be more suitable for consistency reasons, the end user most of the time wont need anything more that the config and the client, so we save some name-space by a little.
Another reason is that aliasing the type allows us to add configuration settings in the future without breaking compatibility.